contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,486 @@
+   present. This is a recommended way of using VPA.
+   
+7. **VPA Updater** is a component responsible for **real-time updates** of Pods.
+   If a Pod uses VPA in `"auto"` mode, the Updater can decide to update it with
+   recommender resources.
+   In MVP this is realized by just evicting the Pod in order to have it
+   recreated with new resources. This approach requires the Pod to belong to a
+   Replica Set (or some other owner capable of recreating it).
+   In future the Updater will take advantage of in-place updates, which would
+   most likely lift this constraint.
+   Because restarting/rescheduling Pods is disruptive to the service, it must be
+   rare.
+   
+8. VPA only controls the resource **request** of containers. It sets the limit
+   to infinity. The request is calculated based on analysis of the current and
+   previous runs (see [Recommendation model](#recommendation-model) below).
+   
+9. **History Storage** is a component that consumes utilization signals and OOMs
+   (same data as the Recommender) from Heapster and stores it persistently.
+   It is used by the Recommender to **initialize its state on startup**.
+   First implementation will be InFluxDB.
+   
+### Architecture overview ###
+![VPA Architecture Diagram](images/vpa-architecture.png "VPA architecture overview")
+
+### API ###
+We introduce a new type of API object `VerticalPodAutoscaler`, which
+consists of the [label selector](https://kubernetes.io/docs/api-reference/v1.5/#labelselector-unversioned)
+for matching Pods and two policy sections: the update policy and the resources
+policy.
+
+#### Example VPA object ####
+```json
+"target": {
+  "matchLabels": [{
+    "key": "app",
+    "value": "my-nginx-1",
+  }]
+}
+"updatePolicy" {
+  "mode": "auto",
+}
+"resourcesPolicy" {
+  "containers": [{
+    "name": "my-nginx",
+    "requests": {
+      "cpu": {
+        "min": "500m",
+      },
+      "memory": {
+        "min": "200Mi",
+      }
+    }
+  }]
+}
+"recommendation" {
+  "containers": [{
+    "name": "my-nginx",
+    "resources": {
+      "requests": {
+        "cpu": "1200m",
+        "memory": "450Mi",
+      }
+    }
+  }]
+}
+```
+
+#### Label Selector ####
+The label selector determines which Pods will be scaled according to the given
+VPA policy. The Recommender will aggregate signals for all Pods matched by a
+given VPA, so it is important that the user set labels to group similarly
+behaving Pods under one VPA.
+
+It is yet to be determined how to resolve conflicts, i.e. when the Pod is
+matched by more than one VPA (this is not a VPA-specific problem though).
+
+#### Update Policy ####
+The update policy controls how VPA applies changes. In MVP it consists of a
+single field `mode` that enables the feature.
+
+```json
+"updatePolicy" {
+  "mode": "",
+}
+```
+
+Mode can be set to one of the following:
+
+1. `"initial"`: VPA only assigns resources on Pod creation and does not
```

Comment:
FYI, API convention for string constants is CamelCase:
https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#constants